dpnp/backend/kernels/dpnp_krnl_random.cpp

```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
+                }
+                for (i = 0; i < size;)
+                {
+                    size_t k, j;
+                    int cv = pvec[idx[i]];
+
+                    // TODO vectorize
+                    for (j = i + 1; (j < size) && (pvec[idx[j]] == cv); j++)
+                    {
+                    }
+                    // assert(j > i);
+                    if (j <= i)
+                    {
+                        throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+                    }
+                    mkl_rng::gamma<_DataType> gamma_distribution(shape + cv, d_zero, d_two);
+                    event_out = mkl_rng::generate(gamma_distribution, DPNP_RNG_ENGINE, j - i, tmp);
+                    event_out.wait();
+
+                    // TODO vectorize
+                    for (k = i; k < j; k++)
+                        result1[idx[k]] = tmp[k - i];
+
+                    i = j;
+                }
+
+                dpnp_memory_free_c(tmp);
+                dpnp_memory_free_c(idx);
+            }
+            else
+            {
+                for (i = 0; i < size; i++)
+                {
+                    mkl_rng::gamma<_DataType> gamma_distribution(shape + pvec[i], d_zero, d_two);
+                    event_out = mkl_rng::generate(gamma_distribution, DPNP_RNG_ENGINE, 1, result1 + 1);
+                    event_out.wait();
+                }
+            }
+            dpnp_memory_free_c(pvec);
+        }
+        else
+        {
+            /* noncentral_chisquare(1, nonc) ~ (Z + sqrt(nonc))**2 for df == 1 */
+            loc = sqrt(nonc);
+            mkl_rng::gaussian<_DataType> gaussian_distribution(loc, d_one);
+            event_out = mkl_rng::generate(gaussian_distribution, DPNP_RNG_ENGINE, size, result1);
+            event_out.wait();
+            event_out = mkl_vm::sqr(DPNP_QUEUE, size, result1, result1, no_deps, mkl_vm::mode::ha);
+            event_out.wait();
+        }
+    }
+    else
+    {
+        double shape, loc;
+        int errcode;
+        size_t i;
+
+        if (df > 1)
+        {
+            double* nvec = nullptr;
+
+            shape = 0.5 * (df - 1.0);
+            /* res has chi^2 with (df - 1) */
+            errcode =
+                vdRngGamma(VSL_RNG_METHOD_GAMMA_GNORM_ACCURATE, get_rng_stream(), size, result1, shape, d_zero, d_two);
+            // TODO
+            // refactor this check, smth like: void status_assert(errcode, VSL_STATUS_OK, func_name)
+            // or just redesign and return int status:
+            // void dpnp_rng_*_c(..., int& status);
+            // or
+            // int dpnp_rng_*_c(...);
+            if (errcode != VSL_STATUS_OK)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+
+            nvec = (double*)mkl_malloc(size * sizeof(double), 64);
```

Comment:
not sure but it looks like it is better to use DPNP function to allocate the memory

Reply:
@shssf no, it is not good. This part of code for condition when `dpnp_queue` is not `cpu`. So for GPU queue we fallback this function implementation on classic MKL, which works only on CPU. So, I think it is bad scenario using DPNP allocator here.

Reply:
You meant it is not SYCL.
hm.. it looks like we need to implement this ability for DPNP allocation functions. To allocate memory with no SYCL queue.

Reply:
I meant that dpnp own allocator work depends on sycl queue and uses instance of global queue. In this part of code queue will be GPU, but classic mkl uses only cpu mem, so this is not right case using dpnp allocator here.

>hm.. it looks like we need to implement this ability for DPNP allocation functions. To allocate memory with no SYCL queue.

yes, this is good point, we can implement it.